code/model/modelreplace.cpp

```diff
@@ -0,0 +1,409 @@
+#include <map>
+#include <memory>
+#include <set>
+#include <vector>
+
+static SCP_unordered_map<SCP_string, std::vector<VirtualPOFDefinition>, SCP_string_lcase_hash, SCP_string_lcase_equal_to> virtual_pofs;
+static SCP_unordered_map<SCP_string, std::function<std::unique_ptr<VirtualPOFOperation>()>> virtual_pof_operations = {
+	{"$Add Subobject:", &make_unique<VirtualPOFOperationAddSubmodel> },
+	{"$Rename Subobjects:", &make_unique<VirtualPOFOperationRenameSubobjects> },
+	{"$Set Subobject Data:", &make_unique<VirtualPOFOperationChangeData> },
+	{"$Set Header Data:", &make_unique<VirtualPOFOperationHeaderData> }
+};
+
+/*
+* forward declares for internal modelread functions
+*/
+extern int read_model_file(polymodel* pm, const char* filename, int ferror, model_read_deferred_tasks& deferredTasks, model_parse_depth depth = {});
+extern void create_family_tree(polymodel* obj);
+extern void model_calc_bound_box(vec3d* box, vec3d* big_mn, vec3d* big_mx);
+
+// General Functions and external code paths
+
+bool model_exists(const SCP_string& filename) {
+	auto it = virtual_pofs.find(filename);
+	if (it != virtual_pofs.end() && !it->second.empty()) {
+		return true;
+	}
+
+	return cf_exists_full(filename.c_str(), CF_TYPE_MODELS);
+}
+
+bool read_virtual_model_file(polymodel* pm, const SCP_string& filename, model_parse_depth depth, int ferror, model_read_deferred_tasks& deferredTasks) {
+	auto virtual_pof_it = virtual_pofs.find(filename);
+
+	//We don't have a virtual pof
+	if(virtual_pof_it == virtual_pofs.end())
+		return false;
+
+	//We have one, but we're already past it and are processing whatever it overwrote
+	int& depthLocal = depth[filename];
+
+	if ((int)virtual_pof_it->second.size() <= depthLocal)
+		return false;
+
+	const auto& virtual_pof = virtual_pof_it->second[depthLocal];
+	depthLocal++;
+
+	read_model_file(pm, filename.c_str(), ferror, deferredTasks, depth);
+
+	for (const auto& operation : virtual_pof.operationList)
+		operation->process(pm, deferredTasks, depth);
+
+	return true;
+}
+
+// Parsing function
+
+static void parse_virtual_pof() {
+	SCP_string name;
+	stuff_string(name, F_NAME);
+
+	auto& virtual_pof_list = virtual_pofs[name];
+	virtual_pof_list.emplace_back();
+	VirtualPOFDefinition& toFill = virtual_pof_list.back();
+
+	toFill.name = name;
+	
+	required_string("+Base POF:");
+	stuff_string(toFill.basePOF, F_FILESPEC);
+
+	do {
+		ignore_white_space();
+		char operation_type[NAME_LENGTH];
+		stuff_string(operation_type, F_NAME, NAME_LENGTH);
+
+		auto operation_parser = virtual_pof_operations.find(operation_type);
+		if (operation_parser != virtual_pof_operations.end())
+			toFill.operationList.emplace_back(operation_parser->second());
+		else {
+			error_display(1, "Unknown operation type %s in virtual POF %s!", operation_type, name.c_str());
+			virtual_pof_list.pop_back();
+			skip_to_start_of_string_either("$POF:", "#End");
+			return;
+		}
+	} while (optional_string_either("$POF:", "#End") == -1);
+	Mp -= 5;
```

Comment:
Is this backing up to the start of `$POF:`?  What happens if it ends the loop on `#End` instead?